processor/resourcedetectionprocessor/internal/resourcedetection.go

```diff
@@ -114,73 +119,113 @@ func NewResourceProvider(logger *zap.Logger, timeout time.Duration, attributesTo
 		timeout:          timeout,
 		detectors:        detectors,
 		attributesToKeep: attributesToKeep,
+		refreshInterval:  0, // No periodic refresh by default
 	}
 }
 
-func (p *ResourceProvider) Get(ctx context.Context, client *http.Client) (resource pcommon.Resource, schemaURL string, err error) {
-	p.once.Do(func() {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, client.Timeout)
-		defer cancel()
-		p.detectResource(ctx, client.Timeout)
-	})
+func (p *ResourceProvider) Get(_ context.Context, _ *http.Client) (pcommon.Resource, string, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	resource := pcommon.NewResource()
+	schemaURL := ""
+	var err error
 
-	return p.detectedResource.resource, p.detectedResource.schemaURL, p.detectedResource.err
+	if p.detectedResource != nil {
+		resource = p.detectedResource.resource
+		schemaURL = p.detectedResource.schemaURL
+		err = p.detectedResource.err
+	}
+
+	return resource, schemaURL, err
 }
 
-func (p *ResourceProvider) detectResource(ctx context.Context, timeout time.Duration) {
-	p.detectedResource = &resourceResult{}
+// Refresh recomputes the resource, replacing any previous result.
+func (p *ResourceProvider) Refresh(ctx context.Context, client *http.Client) (pcommon.Resource, string, error) {
+	ctx, cancel := context.WithTimeout(ctx, client.Timeout)
+	defer cancel()
+
+	res, schemaURL, err := p.detectResource(ctx)
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	prev := p.detectedResource
+
+	// Check if we have a previous successfully snapshot
+	hadPrevSuccess := prev != nil && prev.err == nil && !IsEmptyResource(prev.resource)
```

Comment:
Why the IsEmptyResource case here?  It seems like if a detector returned no error, we should consider that a success, even if it found an empty resource (e.g. isn't running on that cloud or the env var is empty, etc).